File: buffer/cpu/cpp/src/cpu/collection_fun.cpp
#include <collection_fun.h>
#include <cfloat>
#include <algorithm>

enum class Operation {
    Max,
    Min,
    Sum
};

template<Operation Op>
inline void initialize(float* result, int size) {
    if constexpr (Op == Operation::Max) {
        std::fill(result, result + size, -FLT_MAX);
    } else if constexpr (Op == Operation::Min) {
        std::fill(result, result + size, FLT_MAX);
    } else if constexpr (Op == Operation::Sum) {
        std::fill(result, result + size, 0.0f);
    }
}

template<Operation Op>
inline void perform_operation(float& acc, float x) {
    if constexpr (Op == Operation::Max) {
        acc = std::max(acc, x);
    } else if constexpr (Op == Operation::Min) {
        acc = std::min(acc, x);
    } else if constexpr (Op == Operation::Sum) {
        acc += x;
    }
}

template<Operation Op>
inline float reduce_d1(const float* x, size_t size) {
    float acc = x[0];
    for (size_t i = 1; i < size; i++) {
        perform_operation<Op>(acc, x[i]);
    }
    return acc;
}

template<Operation Op>
inline void reduce_d2(const float* x, int xi, int xj, int axis, float* result) {
    if (axis == 0) {
        initialize<Op>(result, xj);
        for (int i = 0; i < xi; i++) {
            const float* x_view = x + i * xj;
            for (int j = 0; j < xj; j++) {
                perform_operation<Op>(result[j], x_view[j]);
            }
        }
    } else if (axis == 1) {
        initialize<Op>(result, xi);
        for (int i = 0; i < xi; i++) {
            const float* x_view = x + i * xj;
            for (int j = 0; j < xj; j++) {
                perform_operation<Op>(result[i], x_view[j]);
            }
        }
    }
}

template<Operation Op>
inline void reduce_d3(const float* x, int xi, int xj, int xk, int axis, float* result) {
    if (axis == 0) {
        initialize<Op>(result, xj * xk);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                float* result_view = result + j * xk;
                const float* x_view = x + (i * xj + j) * xk;
                for (int k = 0; k < xk; k++) {
                    perform_operation<Op>(result_view[k], x_view[k]);
                }
            }
        }
    } else if (axis == 1) {
        initialize<Op>(result, xi * xk);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                float* result_view = result + i * xk;
                const float* x_view = x + (i * xj + j) * xk;
                for (int k = 0; k < xk; k++) {
                    perform_operation<Op>(result_view[k], x_view[k]);
                }
            }
        }
    } else if (axis == 2) {
        initialize<Op>(result, xi * xj);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                float* result_view = result + i * xj + j;
                const float* x_view = x + (i * xj + j) * xk;
                for (int k = 0; k < xk; k++) {
                    perform_operation<Op>(*result_view, x_view[k]);
                }
            }
        }
    }
}

template<Operation Op>
inline void reduce_d4(const float* x, int xi, int xj, int xk, int xl, int axis, float* result) {
    if (axis == 0) {
        initialize<Op>(result, xj * xk * xl);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                for (int k = 0; k < xk; k++) {
                    float* result_view = result + (j * xk + k) * xl;
                    const float* x_view = x + ((i * xj + j) * xk + k) * xl;
                    for (int l = 0; l < xl; l++) {
                        perform_operation<Op>(result_view[l], x_view[l]);
                    }
                }
            }
        }
    } else if (axis == 1) {
        initialize<Op>(result, xi * xk * xl);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                for (int k = 0; k < xk; k++) {
                    float* result_view = result + (i * xk + k) * xl;
                    const float* x_view = x + ((i * xj + j) * xk + k) * xl;
                    for (int l = 0; l < xl; l++) {
                        perform_operation<Op>(result_view[l], x_view[l]);
                    }
                }
            }
        }
    } else if (axis == 2) {
        initialize<Op>(result, xi * xj * xl);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                for (int k = 0; k < xk; k++) {
                    float* result_view = result + (i * xj + j) * xl;
                    const float* x_view = x + ((i * xj + j) * xk + k) * xl;
                    for (int l = 0; l < xl; l++) {
                        perform_operation<Op>(result_view[l], x_view[l]);
                    }
                }
            }
        }
    } else if (axis == 3) {
        initialize<Op>(result, xi * xj * xk);
        for (int i = 0; i < xi; i++) {
            for (int j = 0; j < xj; j++) {
                for (int k = 0; k < xk; k++) {
                    float* result_view = result + (i * xj + j) * xk + k;
                    const float* x_view = x + ((i * xj + j) * xk + k) * xl;
                    for (int l = 0; l < xl; l++) {
                        perform_operation<Op>(*result_view, x_view[l]);
                    }
                }
            }
        }
    }
}

float average_d1(const float* x, size_t size) {
    return reduce_d1<Operation::Sum>(x, size) / size;
}

void average_d2(const float* x, int xi, int xj, int axis, float* result) {
    reduce_d2<Operation::Sum>(x, xi, xj, axis, result);
    if (axis == 0) {
        float inv = 1.0f / xi;
        for (int n = 0; n < xj; n++) {
            result[n] *= inv;
        }
    } else {
        float inv = 1.0f / xj;
        for (int n = 0; n < xi; n++) {
            result[n] *= inv;
        }
    }
}

void average_d3(const float* x, int xi, int xj, int xk, int axis, float* result) {
    reduce_d3<Operation::Sum>(x, xi, xj, xk, axis, result);
    if (axis == 0) {
        float inv = 1.0f / xi;
        for (int n = 0; n < xj * xk; n++) {
            result[n] *= inv;
        }
    } else if (axis == 1) {
        float inv = 1.0f / xj;
        for (int n = 0; n < xi * xk; n++) {
            result[n] *= inv;
        }
    } else {
        float inv = 1.0f / xk;
        for (int n = 0; n < xi * xj; n++) {
            result[n] *= inv;
        }
    }
}

void average_d4(const float* x, int xi, int xj, int xk, int xl, int axis, float* result) {
    reduce_d4<Operation::Sum>(x, xi, xj, xk, xl, axis, result);
    if (axis == 0) {
        float inv = 1.0f / xi;
        for (int n = 0; n < xj * xk * xl; n++) {
            result[n] *= inv;
        }
    } else if (axis == 1) {
        float inv = 1.0f / xj;
        for (int n = 0; n < xi * xk * xl; n++) {
            result[n] *= inv;
        }
    } else if (axis == 2) {
        float inv = 1.0f / xk;
        for (int n = 0; n < xi * xj * xl; n++) {
            result[n] *= inv;
        }
    } else {
        float inv = 1.0f / xl;
        for (int n = 0; n < xi * xj * xk; n++) {
            result[n] *= inv;
        }
    }
}

float max_d1(const float* x, size_t size) {
    return reduce_d1<Operation::Max>(x, size);
}

void max_d2(const float* x, int xi, int xj, int axis, float* result) {
    reduce_d2<Operation::Max>(x, xi, xj, axis, result);
}

void max_d3(const float* x, int xi, int xj, int xk, int axis, float* result) {
    reduce_d3<Operation::Max>(x, xi, xj, xk, axis, result);
}

void max_d4(const float* x, int xi, int xj, int xk, int xl, int axis, float* result) {
    reduce_d4<Operation::Max>(x, xi, xj, xk, xl, axis, result);
}

float min_d1(const float* x, size_t size) {
    return reduce_d1<Operation::Min>(x, size);
}

void min_d2(const float* x, int xi, int xj, int axis, float* result) {
    reduce_d2<Operation::Min>(x, xi, xj, axis, result);
}

void min_d3(const float* x, int xi, int xj, int xk, int axis, float* result) {
    reduce_d3<Operation::Min>(x, xi, xj, xk, axis, result);
}

void min_d4(const float* x, int xi, int xj, int xk, int xl, int axis, float* result) {
    reduce_d4<Operation::Min>(x, xi, xj, xk, xl, axis, result);
}

float sum_d1(const float* x, size_t size) {
    return reduce_d1<Operation::Sum>(x, size);
}

void sum_d2(const float* x, int xi, int xj, int axis, float* result) {
    reduce_d2<Operation::Sum>(x, xi, xj, axis, result);
}

void sum_d3(const float* x, int xi, int xj, int xk, int axis, float* result) {
    reduce_d3<Operation::Sum>(x, xi, xj, xk, axis, result);
}

void sum_d4(const float* x, int xi, int xj, int xk, int xl, int axis, float* result) {
    reduce_d4<Operation::Sum>(x, xi, xj, xk, xl, axis, result);
}

File: buffer/cpu/rs/src/collection_fun.rs
pub fn average_d1(x: &[f32]) -> f32 {
    return sum_d1(x) / x.len() as f32;
}

pub fn average_d2(
    x: &[f32],
    xi: usize, xj: usize,
    axis: usize,
    result: &mut[f32],
) {
    sum_d2(x, xi, xj, axis, result);
    match axis {
        0 => {
            let inv = 1f32 / xi as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        1 => {
            let inv = 1f32 / xj as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}

pub fn average_d3(
    x: &[f32],
    xi: usize, xj: usize, xk: usize,
    axis: usize,
    result: &mut[f32],
) {
    sum_d3(x, xi, xj, xk, axis, result);
    match axis {
        0 => {
            let inv = 1f32 / xi as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        1 => {
            let inv = 1f32 / xj as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        2 => {
            let inv = 1f32 / xk as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}

pub fn average_d4(
    x: &[f32],
    xi: usize, xj: usize, xk: usize, xl: usize,
    axis: usize,
    result: &mut[f32],
) {
    sum_d4(x, xi, xj, xk, xl, axis, result);
    match axis {
        0 => {
            let inv = 1f32 / xi as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        1 => {
            let inv = 1f32 / xj as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        2 => {
            let inv = 1f32 / xk as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        3 => {
            let inv = 1f32 / xl as f32;
            for res in result.iter_mut() {
                *res *= inv;
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}

pub fn max_d1(x: &[f32]) -> f32 {
    return x.iter().fold(f32::MIN, |acc, i| acc.max(*i));
}

pub fn max_d2(
    x: &[f32],
    xi: usize, xj: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MIN);
    reduce_d2(x, xi, xj, axis, result, |acc, i| acc.max(i));
}

pub fn max_d3(
    x: &[f32],
    xi: usize, xj: usize, xk: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MIN);
    reduce_d3(x, xi, xj, xk, axis, result, |acc, i| acc.max(i));
}

pub fn max_d4(
    x: &[f32],
    xi: usize, xj: usize, xk: usize, xl: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MIN);
    reduce_d4(x, xi, xj, xk, xl, axis, result, |acc, i| acc.max(i));
}

pub fn min_d1(x: &[f32]) -> f32 {
    return x.iter().fold(f32::MAX, |acc, i| acc.min(*i));
}

pub fn min_d2(
    x: &[f32],
    xi: usize, xj: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MAX);
    reduce_d2(x, xi, xj, axis, result, |acc, i| acc.min(i));
}

pub fn min_d3(
    x: &[f32],
    xi: usize, xj: usize, xk: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MAX);
    reduce_d3(x, xi, xj, xk, axis, result, |acc, i| acc.min(i));
}

pub fn min_d4(
    x: &[f32],
    xi: usize, xj: usize, xk: usize, xl: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(f32::MAX);
    reduce_d4(x, xi, xj, xk, xl, axis, result, |acc, i| acc.min(i));
}

pub fn sum_d1(x: &[f32]) -> f32 {
    return x.iter().sum();
}

pub fn sum_d2(
    x: &[f32],
    xi: usize, xj: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(0f32);
    reduce_d2(x, xi, xj, axis, result, |acc, i| acc + i);
}

pub fn sum_d3(
    x: &[f32],
    xi: usize, xj: usize, xk: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(0f32);
    reduce_d3(x, xi, xj, xk, axis, result, |acc, i| acc + i);
}

pub fn sum_d4(
    x: &[f32],
    xi: usize, xj: usize, xk: usize, xl: usize,
    axis: usize,
    result: &mut[f32],
) {
    result.fill(0f32);
    reduce_d4(x, xi, xj, xk, xl, axis, result, |acc, i| acc + i);
}

fn reduce_d2<F: Fn(f32, f32) -> f32>(
    x: &[f32],
    xi: usize, xj: usize,
    axis: usize,
    result: &mut[f32],
    block: F,
) {
    assert!(x.len() == xi * xj);
    match axis {
        0 => {
            assert_eq!(result.len(), xj);
            for i in 0..xi {
                let x_i = i * xj;
                for j in 0..xj {
                    result[j] = block(result[j], x[x_i + j]);
                }
            }
        }
        1 => {
            assert_eq!(result.len(), xi);
            for i in 0..xi {
                let x_i = i * xj;
                let mut acc = result[i];
                for j in 0..xj {
                    acc = block(acc, x[x_i + j]);
                }
                result[i] = acc;
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}

fn reduce_d3<F: Fn(f32, f32) -> f32>(
    x: &[f32],
    xi: usize, xj: usize, xk: usize,
    axis: usize,
    result: &mut[f32],
    block: F,
) {
    assert!(x.len() == xi * xj * xk);
    match axis {
        0 => {
            assert_eq!(result.len(), xj * xk);
            for i in 0..xi {
                let x_i = i * xj;
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    let r_i = j * xk; 
                    for k in 0..xk {
                        result[r_i + k] = block(result[r_i + k], x[x_j + k]);
                    }
                }
            }
        }
        1 => {
            assert_eq!(result.len(), xi * xk);
            for i in 0..xi {
                let x_i = i * xj;
                let r_i = i * xk; 
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    for k in 0..xk {
                        result[r_i + k] = block(result[r_i + k], x[x_j + k]);
                    }
                }
            }
        }
        2 => {
            assert_eq!(result.len(), xi * xj);
            for i in 0..xi {
                let x_i = i * xj;
                let r_i = i * xj; 
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    let mut acc = result[r_i + j];
                    for k in 0..xk {
                        acc = block(acc, x[x_j + k]);
                    }
                    result[r_i + j] = acc;
                }
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}

fn reduce_d4<F: Fn(f32, f32) -> f32>(
    x: &[f32],
    xi: usize, xj: usize, xk: usize, xl: usize,
    axis: usize,
    result: &mut[f32],
    block: F,
) {
    assert!(x.len() == xi * xj * xk * xl);
    match axis {
        0 => {
            assert_eq!(result.len(), xj * xk * xl);
            for i in 0..xi {
                let x_i = i * xj;
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    let r_i = j * xk; 
                    for k in 0..xk {
                        let x_k = (x_j + k) * xl;
                        let r_j = (r_i + k) * xl;
                        for l in 0..xl {
                            result[r_j + l] = block(result[r_j + l], x[x_k + l]);
                        }
                    }
                }
            }
        }
        1 => {
            assert_eq!(result.len(), xi * xk * xl);
            for i in 0..xi {
                let x_i = i * xj;
                let r_i = i * xk; 
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    for k in 0..xk {
                        let x_k = (x_j + k) * xl;
                        let r_j = (r_i + k) * xl;
                        for l in 0..xl {
                            result[r_j + l] = block(result[r_j + l], x[x_k + l]);
                        }
                    }
                }
            }
        }
        2 => {
            assert_eq!(result.len(), xi * xj * xl);
            for i in 0..xi {
                let x_i = i * xj;
                let r_i = i * xj; 
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    let r_j = (r_i + j) * xl;
                    for k in 0..xk {
                        let x_k = (x_j + k) * xl;
                        for l in 0..xl {
                            result[r_j + l] = block(result[r_j + l], x[x_k + l]);
                        }
                    }
                }
            }
        }
        3 => {
            assert_eq!(result.len(), xi * xj * xk);
            for i in 0..xi {
                let x_i = i * xj;
                let r_i = i * xj; 
                for j in 0..xj {
                    let x_j = (x_i + j) * xk;
                    let r_j = (r_i + j) * xk;
                    for k in 0..xk {
                        let x_k = (x_j + k) * xl;
                        let mut acc = result[r_j + k];
                        for l in 0..xl {
                            acc = block(acc, x[x_k + l]);
                        }
                        result[r_j + k] = acc;
                    }
                }
            }
        }
        _ => panic!("invalid parameter. [axis: {}]", axis)
    }
}
